spring-cloud-gcp-secretmanager/src/main/java/org/springframework/cloud/gcp/secretmanager/SecretManagerTemplate.java

```diff
@@ -0,0 +1,196 @@
+		return getSecretVersion(secretId, versionName).toByteArray();
+	}
+
+	private ByteString getSecretVersion(String secretId, String versionName) {
+		SecretVersionName secretVersionName = SecretVersionName.of(
+				this.projectIdProvider.getProjectId(),
+				secretId,
+				versionName);
+
+		AccessSecretVersionResponse response = secretManagerServiceClient.accessSecretVersion(secretVersionName);
+
+		return response.getPayload().getData();
+	}
+
+	/**
+	 * Returns the lower-level {@link SecretManagerServiceClient} client object for making API
+	 * calls to Secret Manager service.
+	 *
+	 * <p>
+	 * Useful for executing more advanced use-cases that are not covered by
+	 * {@link SecretManagerTemplate}.
+	 *
+	 * @return the {@link SecretManagerServiceClient} client object.
+	 */
+	public SecretManagerServiceClient getSecretManagerServiceClient() {
+		return this.secretManagerServiceClient;
+	}
+
+	/**
+	 * Create a new version of the secret with the specified payload under a {@link Secret}.
+	 */
+	private void createNewSecretVersion(String secretId, Object rawPayload) {
+		ByteString payload = convertToByteString(rawPayload);
+
+		SecretName name = SecretName.of(projectIdProvider.getProjectId(), secretId);
+		SecretPayload payloadObject = SecretPayload.newBuilder()
+				.setData(payload)
+				.build();
+
+		AddSecretVersionRequest payloadRequest = AddSecretVersionRequest.newBuilder()
+				.setParent(name.toString())
+				.setPayload(payloadObject)
+				.build();
+		secretManagerServiceClient.addSecretVersion(payloadRequest);
+	}
+
+	/**
+	 * Creates a new secret for the GCP Project.
+	 *
+	 * <p>
+	 * Note that the {@link Secret} object does not contain the secret payload. You must
+	 * create versions of the secret which stores the payload of the secret.
+	 */
+	private void createSecret(String secretId) {
+		ProjectName projectName = ProjectName.of(projectIdProvider.getProjectId());
+
+		Secret secret = Secret.newBuilder()
+				.setReplication(
+						Replication.newBuilder()
+								.setAutomatic(Replication.Automatic.newBuilder().build())
+								.build())
+				.build();
+		CreateSecretRequest request = CreateSecretRequest.newBuilder()
+				.setParent(projectName.toString())
+				.setSecretId(secretId)
+				.setSecret(secret)
+				.build();
+		this.secretManagerServiceClient.createSecret(request);
+	}
+
+	/**
+	 * Returns true if there already exists a secret under the GCP project with the
+	 * {@code secretId}.
+	 */
+	private boolean secretExists(String secretId) {
+		ProjectName projectName = ProjectName.of(this.projectIdProvider.getProjectId());
+		ListSecretsPagedResponse listSecretsResponse = this.secretManagerServiceClient.listSecrets(projectName);
+
+		return StreamSupport.stream(listSecretsResponse.iterateAll().spliterator(), false)
+				.anyMatch(secret -> secret.getName().contains(secretId));
+	}
+
+	private ByteString convertToByteString(Object rawPayload) {
```

Comment:
You can get rid of this method if instead you rely on the two possible types: `String` and `byte[]`.

Reply:
Done.